src/ncp/ncp_host.cpp

```diff
@@ -71,6 +75,50 @@ void NcpHost::GetDeviceRole(DeviceRoleHandler aHandler)
     mNcpSpinel.GetDeviceRole(aHandler);
 }
 
+void NcpHost::Join(const otOperationalDatasetTlvs &aActiveOpDatasetTlvs, const AsyncResultReceiver aReceiver)
+{
+    ChainedAsyncResultPtr result;
+    bool                  isAlready    = !mJoinReceivers.empty();
+    auto                  errorHandler = [this](otError aError, const std::string &aErrorInfo) {
+        CallAsyncReceivers(mJoinReceivers, aError, aErrorInfo);
+    };
+
+    mJoinReceivers.push_back(aReceiver);
+    VerifyOrExit(!isAlready);
+
+    result = std::make_shared<ChainedAsyncResult>(errorHandler);
+    result
+        ->First([this, &aActiveOpDatasetTlvs](ChainedAsyncResultPtr aNext) {
+            mNcpSpinel.DatasetSetActiveTlvs(aActiveOpDatasetTlvs, aNext);
+        })
+        ->Then([this](ChainedAsyncResultPtr aNext) { mNcpSpinel.Ip6SetEnabled(true, aNext); })
+        ->Then([this](ChainedAsyncResultPtr aNext) { mNcpSpinel.ThreadSetEnabled(true, aNext); });
```

Comment:
nit: `std::move(aNext)`

Reply:
Not sure if this is needed here `aNext` is a shared_pointer. If we use `std::move`, as I understand, it just destructs the `aNext` within the scope of this lambda function right? I think maybe we should move it when calling `mThen(mNext)` in `SetResult` => `mThen(std::move(mNext))`.  Thoughts?

Reply:
The idea here is to improve the efficiency of argument passing by avoid copy constructions.

There are 2 factors here determining what constructor is triggered when passing arguments:
- The type of the argument to be passed in: Here `aNext` is a non-const value.
- The type of the parameter in the function/method declaration: `NcpSpinel::Ip6SetEnabled` takes an `AsyncTaskPtr` by value.

Hence if you pass `aNext` as is, it will be regarded as an lvalue reference and result in a copy construction of `AsyncTaskPtr` when it's passed to `NcpSpinel::Ip6SetEnabled`. 

However, if you pass in `std::move(aNext)`, it will be passed in as an rvalue reference and result in a move construction, which is a bit cheaper.

Reply:
Oh I see, thanks! Have updated to use `std::move`.